ee/agent/storage/inmemory/keyvalue_store_in_memory.go

```diff
@@ -77,44 +98,68 @@ func (s *inMemoryKeyValueStore) ForEach(fn func(k, v []byte) error) error {
 
 	s.mu.Lock()
 	defer s.mu.Unlock()
-	for k, v := range s.items {
-		if err := fn([]byte(k), v); err != nil {
+	for _, key := range s.order {
+		if err := fn([]byte(key), s.items[key]); err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
+// Update adheres to the Updater interface for bulk replacing data in a key/value store.
+// Note that this method internally defers all mutating operations to the existing Set/Delete
+// functions, so the mutex is not locked here
 func (s *inMemoryKeyValueStore) Update(kvPairs map[string]string) ([]string, error) {
 	if s == nil {
 		return nil, errors.New("store is nil")
 	}
 
-	s.mu.Lock()
-	defer s.mu.Unlock()
-
-	s.items = make(map[string][]byte)
-
 	for key, value := range kvPairs {
 		if key == "" {
 			return nil, errors.New("key is blank")
 		}
 
-		s.items[key] = []byte(value)
+		s.Set([]byte(key), []byte(value))
 	}
 
-	var deletedKeys []string
+	deletedKeys := make([]string, 0)
 
 	for key := range s.items {
 		if _, ok := kvPairs[key]; ok {
 			continue
 		}
 
-		delete(s.items, key)
+		s.Delete([]byte(key))
 
 		// Remember which keys we're deleting
 		deletedKeys = append(deletedKeys, key)
 	}
 
 	return deletedKeys, nil
 }
+
+func (s *inMemoryKeyValueStore) Count() int {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	return len(s.items)
+}
+
+func (s *inMemoryKeyValueStore) AppendValues(values ...[]byte) error {
+	if s == nil {
+		return fmt.Errorf("unable to append values into uninitialized inmemory db store")
+	}
+
+	for _, value := range values {
+		s.Set(s.nextSequenceKey(), value)
+	}
+
+	return nil
+}
+
+func (s *inMemoryKeyValueStore) nextSequenceKey() []byte {
+	s.sequence++
```

Comment:
Do we need a lock in this function in case of simultaneous AppendValues calls...?

Reply:
yeah good call, will add ty